sklearn/decomposition/nmf.py

```diff
@@ -600,9 +740,287 @@ def _fit_coordinate_descent(X, W, H, tol=1e-4, max_iter=200, alpha=0.001,
+
+    H : array-like, shape (n_components, n_features)
+        Initial guess for the solution.
+
+    beta_loss : float or string, default 'frobenius'
+        String must be in {'frobenius', 'kullback-leibler', 'itakura-saito'}.
+        Beta divergence to be minimized, measuring the distance between X
+        and the dot product WH. Note that values different from 'frobenius'
+        (or 2) and 'kullback-leibler' (or 1) lead to significantly slower
+        fits.
+
+    max_iter : integer, default: 200
+        Number of iterations.
+
+    tol : float, default: 1e-4
+        Tolerance of the stopping condition.
+
+    l1_reg_W : double, default: 0.
+        L1 regularization parameter for W.
+
+    l1_reg_H : double, default: 0.
+        L1 regularization parameter for H.
+
+    l2_reg_W : double, default: 0.
+        L2 regularization parameter for W.
+
+    l2_reg_H : double, default: 0.
+        L2 regularization parameter for H.
+
+    update_H : boolean, default: True
+        Set to True, both W and H will be estimated from initial guesses.
+        Set to False, only W will be estimated.
+
+    verbose : integer, default: 0
+        The verbosity level.
+
+    Returns
+    -------
+    W : array, shape (n_samples, n_components)
+        Solution to the non-negative least squares problem.
+
+    H : array, shape (n_components, n_features)
+        Solution to the non-negative least squares problem.
+
+    n_iter : int
+        The number of iterations done by the algorithm.
+
+    References
+    ----------
+    Fevotte, C., & Idier, J. (2011). Algorithms for nonnegative matrix
+    factorization with the beta-divergence. Neural Computation, 23(9).
+    """
+    beta_loss = _beta_loss_to_float(beta_loss)
+
+    # gamma for Maximization-Minimization (MM) algorithm [Fevotte 2011]
+    if beta_loss < 1:
+        gamma = 1. / (2. - beta_loss)
+    elif beta_loss > 2:
+        gamma = 1. / (beta_loss - 1.)
+    else:
+        gamma = 1.
+
+    # used for the convergence criterion
+    previous_error = np.inf
+
+    start_time = time.time()
+    H_sum, HHt, XHt = None, None, None
+    for epoch in range(max_iter):
+        # update W
+        # H_sum, HHt and XHt are saved and reused if not update_H
+        delta_W, H_sum, HHt, XHt = _multiplicative_update_w(
+            X, W, H, beta_loss, l1_reg_W, l2_reg_W, gamma,
+            H_sum, HHt, XHt, update_H)
+        W *= delta_W
+
+        # update H
+        if update_H:
+            delta_H = _multiplicative_update_h(X, W, H, beta_loss, l1_reg_H,
+                                               l2_reg_H, gamma)
+            H *= delta_H
+
+            # These values will be recomputed since H changed
+            H_sum, HHt, XHt = None, None, None
+
+        # test convergence criterion every 100 iterations
+        if tol > 0 and epoch % 100 == 0:
+            error = beta_divergence(X, W, H, beta_loss)
+            if (previous_error - error) / error < tol:
+                break
```

Comment:
Actually this is too strict and never works on some toy problem I am testing on. I thought about storing the initial value of the error in a variable called `error_at_init` and then checking for:

```python
if (previous_error - error) / error_at_init < tol:
    break
```

but then it's hard to define a good value for tol.